pkg/cmd/serve.go

```diff
@@ -50,116 +53,142 @@ func RegisterServeFlags(cmd *cobra.Command, config *server.Config) error {
-	cmd.Flags().StringSliceVar(&config.PresharedSecureKey, PresharedKeyFlag, []string{}, "preshared key(s) to require for authenticated requests")
-	cmd.Flags().DurationVar(&config.ShutdownGracePeriod, "grpc-shutdown-grace-period", 0*time.Second, "amount of time after receiving sigint to continue serving")
-	if err := cmd.MarkFlagRequired(PresharedKeyFlag); err != nil {
+	util.RegisterGRPCServerFlags(grpcFlagSet, &config.GRPCServer, "grpc", "gRPC", ":50051", true)
+	grpcFlagSet.StringSliceVar(&config.PresharedSecureKey, PresharedKeyFlag, []string{}, "preshared key(s) to require for authenticated requests")
+	grpcFlagSet.DurationVar(&config.ShutdownGracePeriod, "grpc-shutdown-grace-period", 0*time.Second, "amount of time after receiving sigint to continue serving")
+	if err := cobra.MarkFlagRequired(grpcFlagSet, PresharedKeyFlag); err != nil {
 		return fmt.Errorf("failed to mark flag as required: %w", err)
 	}
 
+	// Flags for HTTP gateway
+	httpFlags := nfs.FlagSet("HTTP")
+	util.RegisterHTTPServerFlags(httpFlags, &config.HTTPGateway, "http", "gateway", ":8443", false)
+	httpFlags.StringVar(&config.HTTPGatewayUpstreamAddr, "http-upstream-override-addr", "", "Override the upstream to point to a different gRPC server")
+	if err := httpFlags.MarkHidden("http-upstream-override-addr"); err != nil {
+		return fmt.Errorf("failed to mark flag as hidden: %w", err)
+	}
+	httpFlags.StringVar(&config.HTTPGatewayUpstreamTLSCertPath, "http-upstream-override-tls-cert-path", "", "Override the upstream TLS certificate")
+	if err := httpFlags.MarkHidden("http-upstream-override-tls-cert-path"); err != nil {
+		return fmt.Errorf("failed to mark flag as hidden: %w", err)
+	}
+	httpFlags.BoolVar(&config.HTTPGatewayCorsEnabled, "http-cors-enabled", false, "DANGEROUS: Enable CORS on the http gateway")
+	if err := httpFlags.MarkHidden("http-cors-enabled"); err != nil {
+		return fmt.Errorf("failed to mark flag as hidden: %w", err)
+	}
+	httpFlags.StringSliceVar(&config.HTTPGatewayCorsAllowedOrigins, "http-cors-allowed-origins", []string{"*"}, "Set CORS allowed origins for http gateway, defaults to all origins")
+	if err := httpFlags.MarkHidden("http-cors-allowed-origins"); err != nil {
+		return fmt.Errorf("failed to mark flag as hidden: %w", err)
+	}
+
+	apiFlags := nfs.FlagSet("SpiceDB API")
+	// Flags for configuring API behavior
+	// In a future version these will probably be prefixed.
+	apiFlags.Uint16Var(&config.MaximumPreconditionCount, "update-relationships-max-preconditions-per-call", 1000, "maximum number of preconditions allowed for WriteRelationships and DeleteRelationships calls")
+	// TODO: this one should maybe not live in API?
+	apiFlags.Uint64Var(&config.MaxDatastoreReadPageSize, "max-datastore-read-page-size", 1_000, "limit on the maximum page size that we will load into memory from the datastore at one time")
+	apiFlags.BoolVar(&config.DisableV1SchemaAPI, "disable-v1-schema-api", false, "disables the V1 schema API")
```

Comment:
I think we can hide this flag, I don't think there's much utility for most users.